provisioning/device/src/transport_state_machine.ts

```diff
@@ -0,0 +1,253 @@
+                this._fsm.transition('disconnecting', callback, err);
+              } else {
+                this._fsm.transition('connected', callback);
+              }
+            });
+          },
+          '*': () => this._fsm.deferUntilTransition()
+        },
+        connected: {
+          _onEnter: (callback, err, body, result) => {
+            callback(err, body, result);
+          },
+          connect: (callback) => {
+            /* Codes_SRS_NODE_PROVISIONING_TRANSPORT_STATE_MACHINE_18_004: [ If the transport is already connected, then `connect` shall do nothing and call the `callback` immediately. ] */
+            // nothing to do.
+            callback();
+          },
+          disconnect: (callback) => {
+            this._fsm.transition('disconnecting', callback);
+          },
+          register: (callback, registrationId, authorization, requestBody, forceRegistration) => {
+            this._fsm.transition('sending_registration_request', callback, registrationId, authorization, requestBody, forceRegistration);
+          },
+        },
+        sending_registration_request: {
+          _onEnter: (callback, registrationId, authorization, requestBody, forceRegistration) => {
+            /* Codes_SRS_NODE_PROVISIONING_TRANSPORT_STATE_MACHINE_18_012: [ `register` shall call `TransportHandlers.registrationRequest`. ] */
+            this._pendingOperations.operationStarted(callback);
+            this._transport.registrationRequest(registrationId, authorization, requestBody, forceRegistration, (err, body, result, pollingInterval) => {
+              // Check if the operation is still pending before transitioning.  We might be in a different state now and we don't want to mess that up.
+              if (this._pendingOperations.operationIsStillPending(callback)) {
+                this._fsm.transition('response_received', callback, err, registrationId, body, result, pollingInterval);
+              }
+            });
+          },
+          disconnect: (callback) => this._fsm.transition('disconnecting', callback),
+          /* Codes_SRS_NODE_PROVISIONING_TRANSPORT_STATE_MACHINE_18_006: [ If `connect` is called while the transport is executing the first registration request, it shall do nothing and call `callback` immediately. ] */
+          connect: (callback) => callback(),
+          /* Codes_SRS_NODE_PROVISIONING_TRANSPORT_STATE_MACHINE_18_024: [ If `register` is called while a different request is in progress, it shall fail with an `InvalidOperationError`. ] */
+          register: (callback) => callback(new errors.InvalidOperationError('another operation is in progress'))
+        },
+        response_received: {
+          _onEnter: (callback, err, registrationId, body, result, pollingInterval) => {
+            if (err) {
+              /* Codes_SRS_NODE_PROVISIONING_TRANSPORT_STATE_MACHINE_18_013: [ If `TransportHandlers.registrationRequest` fails, `register` shall fail. ] */
```

Comment:
why not have this if(err) in _onEnter of the on sending_registration_request so that you transition directly into the right state?